tortoise/algorithm.go

```diff
@@ -211,6 +211,43 @@ func (t *Tortoise) OnBallot(ballot *types.Ballot) {
 	}
 }
 
+// DecodedBallot created after unwrapping exceptions list and computing internal opinion.
+type DecodedBallot struct {
+	t    *Tortoise
+	info *ballotInfo
+}
+
+// Opinion returns computed opinion.
+func (d *DecodedBallot) Opinion() types.Hash32 {
+	return d.info.opinion()
+}
+
+// CancelWeight zeroes out ballot weight.
+func (d *DecodedBallot) CancelWeight() {
+	d.info.weight = weight{}
+}
+
+// Store adds decoded ballot to the tortoise in-memory state.
+func (d *DecodedBallot) Store() error {
+	d.t.mu.Lock()
+	defer d.t.mu.Unlock()
```

Comment:
this shared lock seems unsafe. mesh.go have access to a `DecodedBallot` and `Tortoise` too. bad code that can cause Tortoise hanging.